provider/pkg/provider/environment.go

```diff
@@ -0,0 +1,271 @@
+	}
+	if inputMap["yaml"].HasValue() && inputMap["yaml"].IsAsset() {
+		input.Yaml = []byte(inputMap["yaml"].AssetValue().Text)
+	} else {
+		return nil, fmt.Errorf("failed to unmarshal yaml value from properties: %s", inputMap)
+	}
+
+	return &input, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Diff(req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
+	olds, err := plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	diffs := olds.Diff(news)
+	if diffs == nil {
+		return &pulumirpc.DiffResponse{
+			Changes: pulumirpc.DiffResponse_DIFF_NONE,
+		}, nil
+	}
+
+	dd := plugin.NewDetailedDiffFromObjectDiff(diffs, false)
+
+	detailedDiffs := map[string]*pulumirpc.PropertyDiff{}
+	replaces := []string(nil)
+	replaceProperties := map[string]bool{
+		"organization": true,
+		"name":         true,
+	}
+	for k, v := range dd {
+		if _, ok := replaceProperties[k]; ok {
+			v.Kind = v.Kind.AsReplace()
+			replaces = append(replaces, k)
+		}
+		detailedDiffs[k] = &pulumirpc.PropertyDiff{
+			Kind:      pulumirpc.PropertyDiff_Kind(v.Kind),
+			InputDiff: v.InputDiff,
+		}
+	}
+
+	changes := pulumirpc.DiffResponse_DIFF_NONE
+	if len(detailedDiffs) > 0 {
+		changes = pulumirpc.DiffResponse_DIFF_SOME
+	}
+	return &pulumirpc.DiffResponse{
+		Changes:             changes,
+		Replaces:            replaces,
+		DetailedDiff:        detailedDiffs,
+		HasDetailedDiff:     true,
+		DeleteBeforeReplace: len(replaces) > 0,
+	}, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Delete(req *pulumirpc.DeleteRequest) (*pbempty.Empty, error) {
+	input, err := ToPulumiServiceEnvironmentInput(req.GetProperties())
+	if err != nil {
+		return nil, err
+	}
+
+	err = st.client.DeleteEnvironment(context.Background(), input.OrgName, input.EnvName)
+	if err != nil {
+		return nil, err
+	}
+	return &pbempty.Empty{}, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Create(req *pulumirpc.CreateRequest) (*pulumirpc.CreateResponse, error) {
+	input, err := ToPulumiServiceEnvironmentInput(req.GetProperties())
+	if err != nil {
+		return nil, err
+	}
+
+	// First check if yaml is valid
+	_, diagnostics, err := st.client.CheckYAMLEnvironment(context.Background(), input.OrgName, input.Yaml)
+	if err != nil {
+		return nil, err
+	}
+	if diagnostics != nil {
+		return nil, fmt.Errorf("failed to create environment, yaml code failed following checks: %+v", diagnostics)
+	}
+
+	// Then create environment, and update it with yaml provided. ESC API architecture doesn't let you do it in one call
+	err = st.client.CreateEnvironment(context.Background(), input.OrgName, input.EnvName)
+	if err != nil {
+		return nil, err
+	}
+	diagnostics, err = st.client.UpdateEnvironment(context.Background(), input.OrgName, input.EnvName, input.Yaml, ETAG)
```

Comment:
Should be able to just pass the empty string instead of `ETAG` here.

Reply:
Replaced with empty string, I re-read ESC client comments and realized I misunderstood the point of ETAG, I thought it was like a source of change, to prevent simultaneous updates.